vbr/vbrCore/functions/sr_visc_calc_BKHK2023.m

```diff
@@ -0,0 +1,115 @@
+function VBR= sr_visc_calc_BKHK2023(VBR)
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%
+% VBR= sr_visc_calc_BKHK2023(VBR)
+% DH 9/8/2025, work in progress 
+%
+% calculates strain rates and viscosities for input state variables.
+%
+% reference:
+% Breithaupt, T., Katz, R. F., Hansen, L. N., & Kumamoto, K. M. (2023). Dislocation theory of steady and transient creep of crystalline solids: Predictions for olivine. Proceedings of the National Academy of Sciences, 120(8), e2203448120.
+%
+% Parameters:
+% -----------
+% VBR   the VBR structure, with state variables in VBR.in.SV. and parameters
+%       loaded in VBR.in.viscous.BKHK2023
+%
+% Ouptut:
+% -------
+% VBR   the VBR structure with new fields
+%       VBR.out.viscous.BKHK2023.sr and .eta.
+%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+% extract state variables and parameters
+T_K = VBR.in.SV.T_K ; % [K], temperature
+sig = VBR.in.SV.sig_MPa.*1e6; % deviatoric stress [Pa]
+d = VBR.in.SV.dg_um ; % [um], grain size
+
+params=VBR.in.viscous.BKHK2023;
+
+% looping over state variables
+for ii = 1:length(T_K)
+    for jj = 1: length(sig)
+        for kk = 1:length(d)
```

Comment:
this should use a single linear index: all of the state variables arrays are required to have the same shape. So what you do is first set up your output arrays so that they'll have the same shape as your inputs:

```
% pre-allocate so that outputs will have the same 
% shapes as inputs
SV_shape = size(T_K);
sr = zeros(SV_shape);
eta = zeros(SV_shape);

% now do a single loop over linear index
n_SVs = numel(T_K);
for i_sv = 1:n_SVS
    % <--- all your calculations --->
    sr(i_sv) = sr result for this thermodynamic state
    eta(i_sv) = eta result for this thermodynamic state
end

% store the outputs in the output structures
VBR.out.viscous.BKHK2023.gbnp.sr = sr;
VBR.out.viscous.BKHK2023.gbnp.eta = eta;
```